clients/src/main/java/org/apache/kafka/common/requests/LeaderAndIsrResponse.java

```diff
@@ -89,7 +90,9 @@ public Errors error() {
 
     @Override
     public Map<Errors, Integer> errorCounts() {
-        return errorCounts(error);
+        if (error != Errors.NONE)
+            return Collections.singletonMap(error, responses.size());
```

Comment:
Is this just a performance optimisation given the code in this PR or does it affect correctness?

Reply:
I was trying to handle the possibility that the top-level error is set after the individual responses. Unfortunately, we can't always use the response errors since, annoyingly, there seems to be a case where we return only the top-level error code in the LeaderAndIsrResponse and no partition responses.

Reply:
Would `responses.size()` be correct in this annoying case?

Reply:
That's a fair point. I guess I could do `max(1, responses.size())`.

Reply:
Sounds good. An alternative would be to change the annoying case to fill the responses. Is that not possible?

Reply:
I've been thinking about it. It's a tad annoying to introduce the extra overhead since it hasn't been needed till now, but the consequence is skewed error counts. An alternative might be to just use the top-level error as is currently done. With the change to `LeaderAndIsrRequest.getErrorResponse`, perhaps that's reasonable?

Reply:
Do we ever have `LeaderAndIsr` where some partitions have an error and others don't?

Reply:
We definitely do. A storage error, for example, would be partition-specific. So I guess finally we just have to fix the weird case in which we skip the partition-level errors. Otherwise there's no way to get accurate counts.

Reply:
I think that's the way to go indeed.

Reply:
Assuming that the `if` check is now an optimization here and in `StopReplicaResponse`, can we please add a brief comment stating that?